src/core/MIGRATION.md

```diff
@@ -604,9 +604,138 @@ It is generally a much greater challenge preparing legacy browser-side code for
 
 To complicate matters further, a significant amount of the business logic in Kibana's client-side code exists inside the `ui/public` directory (aka ui modules), and all of that must be migrated as well. Unlike the server-side code where the order in which you migrated plugins was not particularly important, it's important that UI modules be addressed as soon as possible.
 
-Also unlike the server-side migration, we won't concern ourselves with creating shimmed plugin definitions that then get copied over to complete the migration.
+Because usage of angular and `ui/public` modules varies widely between legacy plugins, there is no "one size fits all" solution to migrating your browser-side code to the new platform. The best place to start is by checking with the platform team to help identify the best migration path for your particular plugin.
 
-### Move UI modules into plugins
+That said, we've seen a series of patterns emerge as teams begin migrating browser code. In practice, most migrations will follow a path that looks something like this:
+
+#### 1. Create a plugin definition file
+
+We've found that doing this right away helps you start thinking about your plugin in terms of lifecycle methods and services, which makes the rest of the migration process feel more natural. It also forces you to identify which actions "kick off" your plugin, since you'll need to execute those when the `setup/start` methods are called.
+
+This definition isn't going to do much for us just yet, but as we get further into the process, we will gradually start returning contracts from our `setup` and `start` methods, while also injecting dependencies as arguments to these methods.
+
+```ts
+// public/plugin.ts
+import { CoreSetup, CoreStart, Plugin } from '../../../../core/public';
+import { FooSetup, FooStart } from '../../../../legacy/core_plugins/foo/public';
+
+/** @public */
+export type DemoSetup = {}
+/** @public */
+export type DemoStart = {}
+
+/** @internal */
+export interface DemoSetupPlugins {
+  foo: FooSetup;
+}
+/** @internal */
+export interface DemoStartPlugins {
+  foo: FooStart;
+}
+
+/** @internal */
+export class DemoPublicPlugin implements Plugin<DemoSetup, DemoStart, DemoSetupPlugins, DemoStartPlugins> {
+  public setup(core: CoreSetup, plugins: DemoSetupPlugins): DemoSetup {
+    // kick off your plugin here...
+    return {};
```

Comment:
I think it could be useful to have one realistic function in each public contract - this would help us write further examples with the plugin.

For example `fetchDemoConfiguration` in `setup` and showDemo in `start`